src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+  pushState: function (propsToPush, callback) {
+    //throw error if propsToPush is undefined or not an object
+    invariantArgCheck(propsToPush);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToPush, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //check that there is only one value being pushed in
+      invariantValueCheck(newPropsAtKey);
+      //push new value into the corresponding array in the current state
+      currentStateAtKey.push(newPropsAtKey);
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //popState FUNCTION: match the deepest keys in the passed in object with the corresponding deepest keys in the current state and pop an item off the array in the corresponding key. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToPush OBJECT: required object argument with deepest keys being a boolean
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  popState: function (propsToPop, callback) {
+    //throw error if propsToPop is undefined or not an object
+    invariantArgCheck(propsToPop);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
+    //build new state object
+    var newState = buildNewState(currentState, propsToPop, function(newState, key, currentStateAtKey, newPropsAtKey) {
+      //check that current state at this key is an array
+      invariantArrayCheck(currentStateAtKey);
+      //pop last value off the corresponding array in the current state
+      currentStateAtKey.pop();
+      newState[key] = currentStateAtKey;
+    });
+    //set state with the updated values and a callback (if provided), triggering a re-render
+    this.setState(newState, callback);
+  },
+
+  //unshiftState FUNCTION: unshift the value of the deepest keys in the passed in object to the corresponding deepest keys in the current state. Throws an error if keys don't match or if the deepest key in current state is not an array
+  //@param propsToPush OBJECT: required object argument with deepest keys being a single value to unshift to the array in the corresponding key in the current state
+  //@param callback FUNCTION: optional callback that will be executed once setState is completed and the component is re-rendered
+  unshiftState: function(propsToUnshift, callback) {
+    //throw error if propsToUnshift is undefined or not an object
+    invariantArgCheck(propsToUnshift);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.